src/extensions/segment-integration.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * Extend Segment with extra PostHog JS functionality. Required for things like Recordings and feature flags to work correctly.
+ *
+ * ### Usage
+ *
+ *  ```js
+ *  // After your standard segment anyalytics install
+ *  analytics.load("GOEDfA21zZTtR7clsBuDvmBKAtAdZ6Np");
+ *
+ *  analytics.ready(() => {
+ *    posthog.init('<posthog-api-key>', {
+ *      capture_pageview: false,
+ *      segment: window.analytics, // NOTE: Be sure to use window.analytics here!
+ *    });
+ *    window.analytics.page();
+ *  })
+ *  ```
+ */
+import { PostHog } from '../posthog-core'
+
+// Loosely based on https://github.com/segmentio/analytics-next/blob/master/packages/core/src/plugins/index.ts
+interface SegmentPluginContext {
+    event: {
+        event: string
+        userId?: string
+        properties: any
+    }
+}
+
+interface SegmentPlugin {
+    name: string
+    version: string
+    type: 'enrichment'
+    isLoaded: () => boolean
+    load: (ctx: SegmentPluginContext, instance: any, config?: any) => Promise<unknown>
+    unload?: (ctx: SegmentPluginContext, instance: any) => Promise<unknown> | unknown
+    ready?: () => Promise<unknown>
+    track?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+    identify?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+    page?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+    group?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+    alias?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+    screen?: (ctx: SegmentPluginContext) => Promise<SegmentPluginContext> | SegmentPluginContext
+}
+
+export const createSegmentIntegration = (posthog: PostHog): SegmentPlugin => {
+    const enrichEvent = (ctx: SegmentPluginContext, eventName: string) => {
+        if (ctx.event.userId) {
+            posthog.register({
```

Comment:
we'll need to reload feature flags here if posthog's distinctID is changing, as otherwise the flags will be incorrect.

Reply:
Which effectively means the first event will not get flag information, but I don't think there's anything we can do about it, unless we can change the model here somehow (not too aware of segment plugins) to also update the ID when segment's identify is called & reload flags.

Reply:
Hmmm is there a nicer way of doing this perhaps? Wanted to do minimal code changes but perhaps I could pull out the distinctId changing code to a proper function...

Reply:
Ooh yes! Just looked at the docs ->

https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#example-plugins

All you'd want to do is call `posthog.identify()` when `analytics.identify()` is called, and everything _should_ be the same, instead of doing `posthog.register` like we're doing here.

Reply:
But won't that end up with double identify calls?

Reply:
Ah, maybe I'm confused about how this integration works.

Is `posthog` the "main" thing used or is `segment` the main thing used on the website when this integration is enabled?

Reply:
or in other words, how is `posthog.identify` called when using the segment integration with the current code?

Reply:
Well if someone is using Segment then they likely want to use Segment I would assume 😅 
Posthog is just added so they can take advantage of the extra features like session recordings. (see https://github.com/PostHog/posthog.com/pull/4671)

In theory though this work should allow them to use posthog stuff like feature flags with underlying additional event tracking still working

Reply:
Right, makes sense.

Ah, I see, the second identify call would occur because one will be coming from segment, and the other one from posthog. Yeah, annoying. 

In that case just checking here, before the register, if `ctx.event.userId != posthog.get_distinct_id()` then reload feature flags should be good enough then.

Reply:
`$feature_flag_called` probably still wouldn't work, until they do `posthog.get_feature_flag()`, but they should be able to do this even with the integration if they want current feature flag information

Reply:
Yeah I think we're saying the same thing. The reason the user would install posthog js is probably to use things like feature flags and I think the way this is implemented means that calling `posthog.getFeatureFlag(...)` will have the right IDs which means any events posthog-js sends should work.